File: Busca/Implementacao/02-Pesquisa Binaria/Python/binary_search.py

# Pesquisa Binária (Binary Search)

def binary_search(arr, target):
    left , right = 0, len(arr) -1 # Listas começam com 0 (zero) 0 a 99
    
    while left <= right:
        mid = (left + right) // 2 #Encontra o meio da lista
        if arr[mid] == target:
            return mid  # Retorna o índice do elemento encontrado
        elif arr[mid] < target: #Busca na metade esquerda
            left = mid + 1 #Faz o left ter um novo valor, eliminando a metade direita
        else:
            right = mid -1 #Faz o right ter um novo valor, eliminando a metade esqueda
    return -1

#Exemplo de Execução
arr = [10, 20, 30, 40, 50, 60, 70]
target = 40
result = binary_search(arr, target)
print(f"Elemento encontrado no índice: {result}" if result != -1 else "Elemento não encontrado")

File: Busca/Implementacao/02-Pesquisa Binaria/Javascript/binary_search.js
// Pesquisa Binária (Binary Search)

function binary_search(arr, target){
    let left = 0, right = arr.length -1;
    
    while(left <= right){
        let mid = Math.floor((left + right) / 2); // A função Math.floor(x) arredonda para baixo o valor de x, retornando o maior número inteiro menor ou igual a x.

        if(arr[mid] === target){
            return mid;
        } else if(arr[mid] < target){
            left = mid +1;
        } else {
            right = mid -1;
        }
        return -1;
    };
};

// Exemplo de Execução:
let arr = [10, 20, 30, 40, 50, 60, 70];
let target = 40;
result = binary_search(arr, target);
console.log(result !== -1 ? `Elemento encontrado no índice: ${result}` : `Elemento não encontrado`);

